File: labs/lingo/lab1/lab1_1.ts
import { Stack } from "../../mathlogic/lab2/lab2_1";

export const BRACKET_PAIRS: Record<string, string> = {
  "(": ")",
  "{": "}",
  "[": "]",
} as const;

export type OpeningBrackets = keyof typeof BRACKET_PAIRS;

export type ClosingBrackets = (typeof BRACKET_PAIRS)[OpeningBrackets];

export class PushdownAutomata<O extends string, C extends string> extends Stack<
  O | C
> {
  private readonly bracketPairs: Record<O, C>;
  private readonly openingBrackets: Set<O>;
  private readonly closingBrackets: Set<C>;

  constructor(pairs: Record<O, C>) {
    super();
    this.bracketPairs = pairs;
    this.openingBrackets = new Set(Object.keys(pairs) as O[]);
    this.closingBrackets = new Set(Object.values(pairs) as C[]);
  }

  private isOpening(char: string): char is O {
    return this.openingBrackets.has(char as O);
  }

  private isClosing(char: string): char is C {
    return this.closingBrackets.has(char as C);
  }

  clean() {
    while (!this.is_empty()) {
      this.pop();
    }
  }

  process(sequence: string) {
    this.clean();
    for (const ch of sequence) {
      if (this.isOpening(ch)) {
        this.push(ch);
        continue;
      }
      if (this.isClosing(ch)) {
        if (this.is_empty()) {
          throw new Error("Wrong sequence!");
        }
        const last = this.pop();
        const expected = this.bracketPairs[last as O];

        if (expected != ch) {
          throw new Error(`Wrong sequence! Expected ${expected}, got ${ch}`);
        }
      }
    }

    if (!this.is_empty()) {
      throw new Error("Wrong sequence! Not all brackets were closed!");
    }

    return true;
  }
}

File: labs/lingo/lab1/lab1_1.py
class PushdownAutomata:
    def __init__(self, pairs: dict):
        self.bracket_pairs = pairs
        self.opening_brackets = set(pairs.keys())
        self.closing_brackets = set(pairs.values())
        self.stack = []

    def clean(self):
        self.stack.clear()

    def process(self, sequence: str) -> bool:
        self.clean()
        for ch in sequence:
            if ch in self.opening_brackets:
                self.stack.append(ch)
                continue
            
            if ch in self.closing_brackets:
                if len(self.stack) == 0:
                    raise Exception("Wrong sequence! (Stack is empty)")
                
                last = self.stack.pop()
                expected = self.bracket_pairs[last]
                
                if expected != ch:
                    raise Exception(f"Wrong sequence! Expected {expected}, got {ch}")
        
        if len(self.stack) > 0:
            raise Exception("Wrong sequence! Not all brackets were closed!")
            
        return True